server/internal/toolsets/impl.go

```diff
@@ -414,6 +414,148 @@ func (s *Service) GetToolset(ctx context.Context, payload *gen.GetToolsetPayload
+	authCtx, ok := contextvalues.GetAuthContext(ctx)
+	if !ok || authCtx == nil || authCtx.ProjectID == nil {
+		return nil, oops.C(oops.CodeUnauthorized)
+	}
+
+	logger := s.logger.With(attr.SlogProjectID(authCtx.ProjectID.String()), attr.SlogToolsetSlug(string(payload.Slug)))
+
+	// Get the original toolset
+	originalToolset, err := s.repo.GetToolset(ctx, repo.GetToolsetParams{
+		Slug:      conv.ToLower(payload.Slug),
+		ProjectID: *authCtx.ProjectID,
+	})
+	if err != nil {
+		return nil, oops.E(oops.CodeNotFound, err, "toolset not found").Log(ctx, logger)
+	}
+
+	// Generate new slug with _copy suffix
+	slugSuffix, err := conv.GenerateRandomSlug(5)
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, err, "failed to generate random slug").Log(ctx, logger)
+	}
+
+	newName := originalToolset.Name + "_copy"
+	newSlug := conv.ToSlug(newName)
+	mcpSlug := authCtx.OrganizationSlug + "-" + slugSuffix
+
+	// Create the cloned toolset
+	clonedToolset, err := s.repo.CreateToolset(ctx, repo.CreateToolsetParams{
+		OrganizationID:         authCtx.ActiveOrganizationID,
+		ProjectID:              *authCtx.ProjectID,
+		Name:                   newName,
+		Slug:                   newSlug,
+		Description:            originalToolset.Description,
+		DefaultEnvironmentSlug: originalToolset.DefaultEnvironmentSlug,
+		HttpToolNames:          originalToolset.HttpToolNames,
+		McpSlug:                conv.ToPGText(mcpSlug),
+		McpEnabled:             false, // Don't auto-enable MCP for cloned toolsets
+	})
+	if err != nil {
+		var pgErr *pgconn.PgError
+		if errors.As(err, &pgErr) && pgErr.Code == pgerrcode.UniqueViolation {
+			// Try with a numbered suffix if _copy already exists
+			for i := 2; i <= 10; i++ {
+				newName = fmt.Sprintf("%s_copy%d", originalToolset.Name, i)
+				newSlug = conv.ToSlug(newName)
+				clonedToolset, err = s.repo.CreateToolset(ctx, repo.CreateToolsetParams{
+					OrganizationID:         authCtx.ActiveOrganizationID,
+					ProjectID:              *authCtx.ProjectID,
+					Name:                   newName,
+					Slug:                   newSlug,
+					Description:            originalToolset.Description,
+					DefaultEnvironmentSlug: originalToolset.DefaultEnvironmentSlug,
+					HttpToolNames:          originalToolset.HttpToolNames,
+					McpSlug:                conv.ToPGText(mcpSlug),
+					McpEnabled:             false,
+				})
+				if err == nil {
+					break
+				}
+				if !errors.As(err, &pgErr) || pgErr.Code != pgerrcode.UniqueViolation {
+					return nil, oops.E(oops.CodeUnexpected, err, "failed to clone toolset").Log(ctx, logger)
+				}
+			}
+			if err != nil {
+				return nil, oops.E(oops.CodeConflict, nil, "could not create unique toolset name").Log(ctx, logger)
+			}
+		} else {
+			return nil, oops.E(oops.CodeUnexpected, err, "failed to clone toolset").Log(ctx, logger)
+		}
+	}
+
+	// Clone toolset version
+	toolURNs := []urn.Tool{}
+	if len(originalToolset.HttpToolNames) > 0 {
```

Comment:
This column isn't the source of truth any more
We should just be copying the latest `toolset_version` directly